### ISSUE TITLE
Upgrade function not running on DB upgrade

### ISSUE THREAD

**Issue by joseprio** [dfahlander/Dexie.js#942]
We noticed some of our users were missing an index that was added through an upgrade. I decided to make a little experiment to see what would happen if the browser "dies" in the middle of an upgrade by using a long `Dexie.waitFor`, and the upgrade function was not called the second time, even though the DB upgrade process was resumed.

I wanted to create a unit test that reproduces this error, but unfortunately I couldn't find a way to express something that doesn't fail nor succeed.

Is this a known problem?

**Comment by dfahlander** [comment]
It's not a known problem, no. If an upgrade transaction fails, it will not upgrade the version, no matter if you were using Dexie.waitFor() or not. However, make sure you return the promise from your upgrader. In case you do not use that, and the Dexie.waitFor() times out (fails with TimeoutError) or stumbles on another error, and that failing promise must be returned to the caller so that it can abort the upgrade transaction. This would be automatically handled if your upgrader is an async function. Otherwise, make sure to return the whole promise flow.

The only known issue regarding upgrades, is that if the initial, database-creating transaction fails, IndexedDB will still create the DB at version 0. This scenario is worked around in Dexie by deleting the database when that happens.

The most common pitfall with Dexie upgrades, is if the version that lacks the index was published at some point, and then adding an index to the same version without incrementing the version number. Dexie won't complain if the installed DB has a different set of indexes than the ones declared.

**Comment by joseprio** [comment]
@dfahlander thanks a lot for your quick reply.

I tweaked one of your jsfiddle, and I can reproduce the problem in Firefox 48 and 58, but not in Firefox 69 or newest Chrome, using Dexie 2.0.4 .
```
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
db.version(1).stores({
	friends: '++id,name,age'
});
db.version(2).upgrade(tx => Dexie.waitFor(sleep(800000)));
```
Could it be a browser bug?

**Comment by dfahlander** [comment]
Was the result that the DB was on version 2 in the failed results? If so, you might have nailed down the issue. But I would like to nail it down further. 

* If you provide a shorter timeout than the default of 60 seconds to [Dexie.waitFor()](https://dexie.org/docs/Dexie/Dexie.waitFor()), and pass a lesser value to the sleep() function, is it then also failing on Firefox 48 or 58?

* If you provide an alternate set of indexes in version 2 and reproduce the problem, what would be the set of indexes and version number present after reproducing it?

**Comment by joseprio** [comment]
Yes, the result is the DB having the updated version; adding an additional index will also apply properly; however, we know that some of our users don't have that index created, so I'm assuming that there's some situation in which things break.
Lowering the values in `Dexie.waitFor` and `sleep` doesn't make a difference, as I refresh the browser way before that.
Interestingly, if I use "Run", refreshing only the frame, it will run properly, so it only fails when I close or refresh the tab.

**Comment by dfahlander** [comment]
I see! I think I missed the detail in your earlier message that you reloaded the browser window. It seems pretty much that it could be a browser bug. The expected behavior would be the system should abort the transaction if cancelled outside of application code.

I suppose we can assume this has been an issue in Firefox that has been solved at some time between 58 and 69. I suppose it would be possible to reproduce the issue that indexes aren't being created if you'd add a version 3 with new indexes. DB would end up on version 3 in this case, on the affected browsers without the created indexes, I suppose.

It would be interesting to verify if this could be an issue on other browsers as well, specifically Safari and IE/Edge. Even if not using Dexie.waitFor(), long running upgrade transaction could be an issue when migrating large databases as well. Hopefully this is an issue only in old firefox browsers and if so, we could leave it behind.

In your particular case:
* Does your app have long-running upgrade transactions between some versions?
* If so, could you solve the upgrade differently? One way would be to subscribe to  the [ready](https://dexie.org/docs/Dexie/Dexie.on.ready) event similar to how it's done in [this sample](https://github.com/dfahlander/Dexie.js/blob/v3.0.0-rc.1/samples/ajax-populate/populateFromAjaxCall.html#L59-L96)?
* To make affected clients' schemas back in sync, I think you'd have to do a workaround by attaching another version with a simpler upgrade function `upgrade(tx)` that does the following:
  1. Use the IDBTransaction from [tx.idbtrans](https://dexie.org/docs/Transaction/Transaction#idbtrans--idbtransaction) to enumerate the possibly missing indexes `tx.idbtrans.objectStore(tableName).indexNames.contains('possiblyMissingIndex')`. If not, use tx.idbtrans.objectStore(tableName).createIndex() on that upgrade transaction to create it. These are synchronous functions.

It would be better if Dexie compared the actual installed indexes instead of relying on the declared schema in upgrades. I hope to do that improvement in a coming version (pull requests welcome ;)) If that would have been the case now, the workaround to get failed clients's schemas in sync would be to just increment the version number. Still it wouldn't be waterproof in the affected browsers, but it would at least repair most of the affected ones.

**Comment by dfahlander** [comment]
Thanks a lot!
* Have you ever stumbled on the issue in any other browser than Firefox version <= 58?
* When it fails, is it failing the same way every time, or is it sporadic?

**Comment by joseprio** [comment]
Will check and let you know.

The Firefox 48 error is always reproducible

**Comment by joseprio** [comment]
Unfortunately I haven't verified if the issue is present in other older browsers (but will check when possible).
I've created a pull request with some changes that should help resolve the issue I described (hopefully, still need to test it); it also has the added bonus that if the older definition is accidentally removed, the upgrade will be able to run and complete successfully. It's an early version, but I decided to put it up to see to get some feedback on the solution.

**Comment by joseprio** [comment]
I had a problem with one unit tests; turns out there was an inconsistency on how `adjustToExistingIndexNames` operates, but probably that information wasn't being used until now.
